### ISSUE TITLE
Sockets.js afterDisconnect being called after Sails.lower

### ISSUE THREAD

**Issue by wedneyyuri** [balderdashy/sails#3397]
Reopening issue #3332 closed by @sailsbot :

**Sails version**: 0.11.2 and 0.11.3

**Problem**
If you do any database manipulation using the `afterDisconnect` handler in [sockets.js](http://sailsjs.org/documentation/anatomy/my-app/config/sockets-js) you will get an error when the Sails.lower() is called.

Ex:
```
  afterDisconnect: function(session, socket, cb) {
    var socketID = sails.sockets.id(socket);
    if (!socketID)
      return cb(new Error('socketID is null!'));

    User.removeSocket(socketID)
    .then(function() {
      cb();
    }, cb);
  },
```
After calling `sails.lower(done);` in [bootstrap.test.js](http://sailsjs.org/documentation/concepts/testing#/bootstrap-test-js) I'm getting this error:
```
  1)  "after all" hook:
     Uncaught TypeError: Cannot read property 'update' of undefined
      at Object.module.exports.adapter.update (C:\Users\wedne_000\Documents\Estudo\rbk-server\node_modules\sails-memory\lib\adapter.js:125:27)
      at module.exports.update (C:\Users\wedne_000\Documents\Estudo\rbk-server\node_modules\sails\node_modules\waterline\lib\waterline\adapter\dql.js:241:13)
      at bound.updateRecords (C:\Users\wedne_000\Documents\Estudo\rbk-server\node_modules\sails\node_modules\waterline\lib\waterline\query\dql\update.js:201:16)
      at C:\Users\wedne_000\Documents\Estudo\rbk-server\node_modules\sails\node_modules\waterline\lib\waterline\query\dql\update.js:52:21
```

**Workaround:**
So, after some debugging, I found a private property (`_exiting`) in `node_modules\sails\lib\app\lower.js`. My current workaround is:
```
  afterDisconnect: function(session, socket, cb) {
    // see node_modules\sails\lib\app\lower.js:32
    // if sails is exiting we can't do database operations
    if (sails._exiting === true) {
      return cb();
    }

    var socketID = sails.sockets.id(socket);
    if (!socketID) {
      return cb(new Error('socketID is null!'));
    }

    User.removeSocket(socketID)
    .then(function() {
      cb();
    }, cb);
  },
```
* Is this the **expected behavior**?
* Is there **any better way** to do this?

**Comment by wedneyyuri** [comment]
Error log:
```
  1)  "after all" hook:
     Uncaught TypeError: Cannot read property 'update' of undefined
      at Object.module.exports.adapter.update (node_modules/sails-disk/lib/adapter.js:140:27)
      at module.exports.update (node_modules/sails/node_modules/waterline/lib/waterline/adapter/dql.js:241:13)
      at bound.updateRecords (node_modules/sails/node_modules/waterline/lib/waterline/query/dql/update.js:207:16)
      at node_modules/sails/node_modules/waterline/lib/waterline/query/dql/update.js:58:21
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:708:13
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:49:16
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:263:32
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:40:16
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:705:17
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:49:16
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:263:32
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:40:16
      at bound.<anonymous> (node_modules/sails/node_modules/waterline/lib/waterline/utils/schema.js:151:44)
      at fn (node_modules/sails/node_modules/waterline/lib/waterline/utils/callbacksRunner.js:79:10)
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:162:20
      at iterate (node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:256:13)
      at Object.async.forEachOfSeries.async.eachOfSeries (node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:275:9)
      at Object.async.forEachSeries.async.eachSeries (node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:211:22)
      at Object.runner.beforeUpdate (node_modules/sails/node_modules/waterline/lib/waterline/utils/callbacksRunner.js:82:9)
      at node_modules/sails/node_modules/waterline/lib/waterline/query/dql/update.js:175:17
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:699:13
      at iterate (node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:256:13)
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:268:29
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:40:16
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:705:17
      at node_modules/sails/node_modules/waterline/lib/waterline/query/validate.js:72:7
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:708:13
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:49:16
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:263:32
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:40:16
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:705:17
      at node_modules/sails/node_modules/waterline/lib/waterline/query/validate.js:66:11
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:49:16
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:263:32
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:40:16
      at node_modules/sails/node_modules/waterline/lib/waterline/query/validate.js:60:13
      at node_modules/sails/node_modules/waterline/lib/waterline/utils/schema.js:151:44
      at runner (node_modules/sails/node_modules/waterline/lib/waterline/query/validate.js:58:11)
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:162:20
      at iterate (node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:256:13)
      at Object.async.forEachOfSeries.async.eachOfSeries (node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:275:9)
      at Object.async.forEachSeries.async.eachSeries (node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:211:22)
      at node_modules/sails/node_modules/waterline/lib/waterline/query/validate.js:64:15
      at node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:699:13
      at iterate (node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:256:13)
```

**Comment by sgress454** [comment]
Your workaround seems fine; another option that doesn't involve a (semi) private property would be to listen for the `lower` event in your bootstrap and clear the `afterDisconnect` handler there:

```
sails.on('lower', function() {delete sails.config.afterDisconnect;}
```

In the end, this probably isn't really a real-world issue; it comes up during testing because you're in control of the sockets _and_ the server at the same time.  Most of the time when you have a Sails instance running in production, if it shuts down for any reason then `afterDisconnect` won't be called because the clients wouldn't even have a chance to disconnect gracefully first.

**Comment by wedneyyuri** [comment]
@sgress454 Please, **reopen** this issue.

In my opinion, we can listen for the `lower` event inside the sails lib as probably this is a real-world issue. Deleting the afterDisconnect handler or accessing a (semi) private can be a bit confusing.

Also, it is happening in **production** every time I turn off the server (*without the workaround*):
![image](https://cloud.githubusercontent.com/assets/7511692/11761977/078a678a-a0bd-11e5-8405-d9de66d1b5b9.png)

**Comment by sgress454** [comment]
@wedneyyuri you're right, I confirmed this locally.  Seems to be a timing issue, I'm taking a look.

**Comment by sgress454** [comment]
Ok, this turned out to be an issue in sails-hook-sockets where the lifecycle callbacks (beforeConnect and afterDisconnect) were being called for the admin socket.  Ignoring the callbacks for the admin socket fixed the issue. Published sails-hook-sockets 0.11.28; you should be able to get it with `npm update sails`.  Thanks for reaching out!
